src/inferencer/util.ts

```diff
@@ -0,0 +1,153 @@
+import * as t from './types'
+import { Scope } from './types'
+
+import { Context, Type } from '../types'
+
+import * as create from '../utils/typeAstCreator'
+
+function numberXnumberRnumber() {
+  // const num = new t.Number()
+  const num = new t.Float()
+  return new t.Function([num, num], num)
+}
+
+function numberXnumberRfloat() {
+  // const num = new t.Number()
+  const num = new t.Float()
+  return new t.Function([num, num], new t.Float())
+}
+
+function comparableXcomparableRcomparable() {
+  const comparable = new t.Comparable()
+  return new t.Function([comparable, comparable], comparable)
+}
+
+function comparableXcomparableRBool() {
+  const comparable = new t.Comparable()
+  return new t.Function([comparable, comparable], new t.Bool())
+}
+
+export const UNARY_MINUS_OPERATOR = 'unary-'
+
+/**
+ * Creates an initial environment that contains types for primitive operators.
+ * @returns The initial environment.
+ */
+export function createEnv(context: Context): t.Environment {
+  const env = new Scope(false)
+
+  // logical operators
+  let type = new t.Var()
+  env.set('||', new t.Function([new t.Bool(), type], type))
+  type = new t.Var()
+  env.set('&&', new t.Function([new t.Bool(), type], type))
+
+  // unary operators
+  env.set('!', new t.Function([new t.Bool()], new t.Bool()))
+  env.set(UNARY_MINUS_OPERATOR, new t.Function([new t.Float()], new t.Float()))
+
+  // binary operators
+  env.set('+', comparableXcomparableRcomparable())
+  env.set('-', numberXnumberRnumber())
+  env.set('*', numberXnumberRnumber())
+  env.set('/', numberXnumberRfloat())
+  env.set('%', numberXnumberRnumber())
+
+  env.set('===', comparableXcomparableRBool())
+  env.set('!==', comparableXcomparableRBool())
+  env.set('<', comparableXcomparableRBool())
+  env.set('<=', comparableXcomparableRBool())
+  env.set('>', comparableXcomparableRBool())
+  env.set('>=', comparableXcomparableRBool())
+
+  let headType = new t.Var()
+  let tailType = new t.Var()
```

Comment:
Since you treat all functions are polymorphic, and fresh variables are generated, is there a need for new vars for each pair/list functions?

Reply:
Yeah, thanks for catching!